test/api/model/model_test.dart

```diff
@@ -212,4 +215,111 @@ void main() {
         .deepEquals([2, 3, 11]);
     });
   });
+
+  group('MessageEditState', () {
+    Map<String, dynamic> baseJson() => deepToJson(eg.streamMessage()) as Map<String, dynamic>;
+
+    group('Edit history is absent', () {
+      test('Message with no evidence of an edit history -> none', () {
+        check(Message.fromJson(baseJson()..['edit_history'] = null))
+          .editState.equals(MessageEditState.none);
+      });
+
+      test('Message without edit history has last edit timestamp -> edited', () {
+        check(Message.fromJson(baseJson()
+            ..['edit_history'] = null
+            ..['last_edit_timestamp'] = 1678139636))
+          .editState.equals(MessageEditState.edited);
+      });
+    });
+
+    void checkEditState(MessageEditState editState, List<Map<String, dynamic>> editHistory){
+      check(Message.fromJson(baseJson()..['edit_history'] = editHistory))
+        .editState.equals(editState);
+    }
+
+    group('edit history exists', () {
+      test('Moved message has last edit timestamp but no actual edits -> moved', () {
+        check(Message.fromJson(baseJson()
+            ..['edit_history'] = [{'prev_stream': 5, 'stream': 7}]
+            ..['last_edit_timestamp'] = 1678139636))
+          .editState.equals(MessageEditState.moved);
+      });
+
+      test('Channel change only -> moved', () {
+        checkEditState(MessageEditState.moved,
+          [{'prev_stream': 5, 'stream': 7}]);
+      });
+
+      test('Topic name change only -> moved', () {
+        checkEditState(MessageEditState.moved,
+          [{'prev_topic': 'old_topic', 'topic': 'new_topic'}]);
+      });
+
+      test('Both topic and content changed -> edited', () {
+        checkEditState(MessageEditState.edited, [
+          {'prev_topic': 'old_topic', 'topic': 'new_topic'},
+          {'prev_content': 'old_content'},
+        ]);
+        checkEditState(MessageEditState.edited, [
+          {'prev_content': 'old_content'},
+          {'prev_topic': 'old_topic', 'topic': 'new_topic'},
+        ]);
+      });
+
+      test('Both topic and content changed in a single edit -> edited', () {
+        checkEditState(MessageEditState.edited,
+          [{'prev_topic': 'old_topic', 'topic': 'new_topic', 'prev_content': 'old_content'}]);
+      });
+
+      test('Content change only -> edited', () {
+        checkEditState(MessageEditState.edited,
+          [{'prev_content': 'old_content'}]);
+      });
+
+      test("'prev_topic' present without the 'topic' field -> moved", () {
+        checkEditState(MessageEditState.moved,
+          [{'prev_topic': 'old_topic'}]);
+      });
+
+      // TODO(server-5) prev_subject was the old name of prev_topic on pre-5.0 servers
```

Comment:
The purpose of a TODO-server comment is to make sure we don't miss a spot where we can delete or simplify code when we drop support for older server versions.

As a result, it's uncommon for tests to need TODO-server comments. The functionality being tested will have its own such comment… and then when we remove that code, normally the point of the test is that it should detect we've done that :slightly_smiling_face: